pages/admin/students.js

```diff
@@ -0,0 +1,236 @@
+import { reformatEmail } from "../../utils/email";
+
+export const getServerSideProps = async ({ req, res }) => {
+  const ssr = await createRequiredAuth({ roles: ["admin"] })({ req, res });
+
+  ssr.props.initialData = (await getStudents()).map(serializeDocument);
+
+  return ssr;
+};
+
+function getColumnsWithActions(actionsFn) {
+  return [
+    {
+      dataField: "fname",
+      text: "First Name",
+    },
+    {
+      dataField: "lname",
+      text: "Last Name",
+    },
+    {
+      dataField: "permNum",
+      text: "Perm",
+    },
+    {
+      dataField: "email",
+      text: "Email",
+    },
+    {
+      dataField: "section",
+      text: "Section",
+    },
+    {
+      dataField: "df1",
+      isDummyField: true,
+      text: "Actions",
+      formatter: actionsFn,
+    },
+  ];
+}
+
+export default function ManageStudentsPage(props) {
+  const { user, initialData } = props;
+  const { showToast } = useToasts();
+  const { data, mutate } = useSWR("/api/students", { initialData });
+  const [newStudentEmail, setNewStudentEmail] = useState("");
+  const [newStudentSection, setNewStudentSection] = useState("");
+  const [newStudentFname, setNewStudentFname] = useState("");
+  const [newStudentLname, setNewStudentLname] = useState("");
+  const [newStudentPermNum, setNewStudentPermNum] = useState("");
+  const addStudent = useCallback(
+    async (e) => {
+      // override default form submission behavior
+      e.preventDefault();
+      e.stopPropagation();
+
+      setNewStudentEmail("");
+      setNewStudentSection("");
+      setNewStudentFname("");
+      setNewStudentLname("");
+      setNewStudentPermNum("");
+      showToast(`Added student ${reformatEmail(newStudentEmail)}`);
+      await mutate(
+        [
+          ...data,
+          {
+            email: reformatEmail(newStudentEmail),
+            section: newStudentSection,
+            fname: newStudentFname,
+            lname: newStudentLname,
+            permNum: newStudentPermNum,
+          },
+        ],
+        false
+      );
+      await fetch("/api/students", {
+        method: "POST",
+        headers: {
+          "Content-Type": "application/json",
+        },
+        body: JSON.stringify({
+          email: reformatEmail(newStudentEmail),
+          section: newStudentSection,
+          fname: newStudentFname,
+          lname: newStudentLname,
+          permNum: newStudentPermNum,
+        }),
+      });
+      await mutate();
+    },
+    [
+      newStudentEmail,
+      newStudentSection,
+      newStudentFname,
+      newStudentLname,
+      newStudentPermNum,
+    ]
+  );
+
+  const deleteStudent = useCallback(async (studentId) => {
+    console.log(studentId);
+    showToast(
+      `Deleted student ${data.find((u) => u._id === studentId)?.email}`
+    );
+    await mutate(
+      data.filter((u) => u._id !== studentId),
+      false
+    );
+    await fetch(`/api/students/${studentId}`, { method: "DELETE" });
+    await mutate();
```

Comment:
Is there any particular reason we have this trailing mutate here? It doesn't appear to be updating any cached state.

Reply:
@Ulramok that's a question for you

Reply:
I believe that tells swr to refetch the data. We used a similar flow in admins.js

Reply:
@btk5h can you weigh in on this? My understanding is that you need to pass `mutate` the path to the cached data as well as the cached data you wish to update; I'm really not sure how that works when nothing is passed to `mutate`.

Reply:
That would be the case with the global `mutate` function. The `mutate` being called here is created by the `useSWR` hook, and is therefore bound to the key we passed to `useSWR`.

Reply:
So then what is the purpose of the call to mutate right before? It seems like we're speculating on the outcome and setting that data in mutate, calling the API, then using mutate to double check that our original speculation was correct? Does that sound right?

Reply:
Right, we want to update the UI as soon as possible to give feedback to the user, but we also want to provide the authoritative representation of the data once it becomes available.